pgmax/groups/variables.py

```diff
@@ -1,73 +1,136 @@
 """A module containing the variables group classes inheriting from the base VariableGroup."""
 
-import collections
-import itertools
 from dataclasses import dataclass
-from typing import Any, Dict, Hashable, Mapping, OrderedDict, Tuple, Union
+from typing import Any, Dict, Hashable, List, Mapping, Tuple, Union
 
 import jax
 import jax.numpy as jnp
 import numpy as np
 
-from pgmax.fg import groups, nodes
+from pgmax.fg import groups
+from pgmax.utils import cached_property
+
+MAX_SIZE = 1e9
 
 
 @dataclass(frozen=True, eq=False)
 class NDVariableArray(groups.VariableGroup):
     """Subclass of VariableGroup for n-dimensional grids of variables.
 
     Args:
-        num_states: The size of the variables in this variable group
-        shape: a tuple specifying the size of each dimension of the grid (similar to
+        shape: Tuple specifying the size of each dimension of the grid (similar to
             the notion of a NumPy ndarray shape)
+        num_states: An integer or an array specifying the number of states of the
+            variables in this VariableGroup
     """
 
-    num_states: int
     shape: Tuple[int, ...]
+    num_states: Union[int, np.ndarray]
+
+    def __post_init__(self):
+        if np.prod(self.shape) > MAX_SIZE:
+            raise ValueError(
+                f"Currently only support NDVariableArray of size smaller than {int(MAX_SIZE)}. Got {np.prod(self.shape)}"
+            )
+
+        if np.isscalar(self.num_states):
+            num_states = np.full(self.shape, fill_value=self.num_states, dtype=np.int64)
+            object.__setattr__(self, "num_states", num_states)
+        elif isinstance(self.num_states, np.ndarray) and np.issubdtype(
+            self.num_states.dtype, int
+        ):
+            if self.num_states.shape != self.shape:
+                raise ValueError(
+                    f"Expected num_states shape {self.shape}. Got {self.num_states.shape}."
+                )
+        else:
+            raise ValueError(
+                "num_states should be an integer or a NumPy array of dtype int"
+            )
 
-    def _get_names_to_variables(
-        self,
-    ) -> OrderedDict[Union[int, Tuple[int, ...]], nodes.Variable]:
-        """Function that generates a dictionary mapping names to variables.
+        # Only compute the hash once, which is guaranteed to be an int64
+        this_id = id(self) % 2**32
+        this_hash = this_id * int(MAX_SIZE)
+        assert this_hash < 2**63
```

Comment:
Not sure about this assert...Given you are doing the `%`, I assume this should never fail? 

A separate question is, although there won't be collisions for `id(self)`, is it possible that there will be collisions for `id(self) % 2**32`? Why are we doing the `%` here?

Reply:
to make sure that the hash is int64